File: packet_seaquence/log/go_src/models/protocol_analyzer/tcp.go
// Package protocol_analyzer provides TCP protocol analyzer
package protocol_analyzer

import (
	"fmt"
	"strings"
)

// AnalyzeTCP implements the ProtocolAnalyzer interface for TCP
type AnalyzeTCP struct {
	BaseProtocolAnalyzer
}

// NewAnalyzeTCP creates a new TCP analyzer
func NewAnalyzeTCP(packet Packet) *AnalyzeTCP {
	return &AnalyzeTCP{
		BaseProtocolAnalyzer: NewBaseProtocolAnalyzer(packet),
	}
}

// Analyze extracts TCP information from a packet
func (a *AnalyzeTCP) Analyze() map[string]interface{} {
	if !a.Pkt.LayerExists("tcp") {
		return nil
	}

	tcpInfo := make(map[string]interface{})
	
	// Source port
	if srcPort, ok := a.Pkt.GetField("tcp", "srcport"); ok {
		tcpInfo["src_port"] = srcPort
	}
	
	// Destination port
	if dstPort, ok := a.Pkt.GetField("tcp", "dstport"); ok {
		tcpInfo["dst_port"] = dstPort
	}
	
	// Sequence number
	if seq, ok := a.Pkt.GetField("tcp", "seq"); ok {
		tcpInfo["seq"] = seq
		if seqRaw, ok := a.Pkt.GetField("tcp", "seq_raw"); ok {
			tcpInfo["seq_raw"] = seqRaw
		}
	}
	
	// Acknowledgment number
	if ack, ok := a.Pkt.GetField("tcp", "ack"); ok {
		tcpInfo["ack"] = ack
		if ackRaw, ok := a.Pkt.GetField("tcp", "ack_raw"); ok {
			tcpInfo["ack_raw"] = ackRaw
		}
	}
	
	// Flags
	if flags, ok := a.Pkt.GetField("tcp", "flags"); ok {
		tcpInfo["flags"] = flags
		tcpInfo["flags_desc"] = a.getTcpFlagsDesc(fmt.Sprintf("%v", flags))
	}
	
	// Window size
	if windowSize, ok := a.Pkt.GetField("tcp", "window_size"); ok {
		tcpInfo["window_size"] = windowSize
	} else if window, ok := a.Pkt.GetField("tcp", "window"); ok {
		tcpInfo["window_size"] = window
	}
	
	// Checksum
	if checksum, ok := a.Pkt.GetField("tcp", "checksum"); ok {
		tcpInfo["checksum"] = checksum
	}
	
	// Urgent pointer
	if urgentPointer, ok := a.Pkt.GetField("tcp", "urgent_pointer"); ok {
		tcpInfo["urgent_pointer"] = urgentPointer
	}
	
	// Options
	if options, ok := a.Pkt.GetField("tcp", "options"); ok {
		tcpInfo["options"] = options
	}
	
	// Payload
	if payload, ok := a.Pkt.GetField("tcp", "payload"); ok {
		tcpInfo["payload"] = payload
	} else if payloadRaw, ok := a.Pkt.GetField("tcp", "payload_raw"); ok {
		tcpInfo["payload_raw"] = payloadRaw
	}
	
	// Stream number
	if stream, ok := a.Pkt.GetField("tcp", "stream"); ok {
		tcpInfo["stream"] = stream
	}
	
	return tcpInfo
}

// GetDisplayInfo returns formatted TCP information
func (a *AnalyzeTCP) GetDisplayInfo() string {
	info := a.Analyze()
	if info == nil {
		return "TCP情報なし"
	}
	
	srcPort := getOrDefault(info["src_port"], "不明")
	dstPort := getOrDefault(info["dst_port"], "不明")
	flags := getOrDefault(info["flags_desc"], "不明")
	seq := getOrDefault(info["seq"], "N/A")
	ack := getOrDefault(info["ack"], "N/A")
	
	portInfo := fmt.Sprintf("%s -> %s", srcPort, dstPort)
	flagInfo := fmt.Sprintf("[%s]", flags)
	
	if strings.Contains(flags, "SYN") && !strings.Contains(flags, "ACK") {
		return fmt.Sprintf("TCP %s %s 接続開始", portInfo, flagInfo)
	} else if strings.Contains(flags, "SYN") && strings.Contains(flags, "ACK") {
		return fmt.Sprintf("TCP %s %s 接続確認", portInfo, flagInfo)
	} else if strings.Contains(flags, "FIN") {
		return fmt.Sprintf("TCP %s %s 接続終了", portInfo, flagInfo)
	} else if strings.Contains(flags, "RST") {
		return fmt.Sprintf("TCP %s %s 接続リセット", portInfo, flagInfo)
	} else if strings.Contains(flags, "ACK") {
		return fmt.Sprintf("TCP %s %s 確認応答 (SEQ=%s, ACK=%s)", portInfo, flagInfo, seq, ack)
	} else {
		return fmt.Sprintf("TCP %s %s", portInfo, flagInfo)
	}
}

// GetSummary returns a summary of TCP information
func (a *AnalyzeTCP) GetSummary() string {
	info := a.Analyze()
	if info == nil {
		return "TCP"
	}
	
	srcPort := getOrDefault(info["src_port"], "")
	dstPort := getOrDefault(info["dst_port"], "")
	flags := getOrDefault(info["flags_desc"], "")
	
	serviceInfo := a.getServiceNameFromPort(dstPort)
	
	if serviceInfo != "" {
		return fmt.Sprintf("TCP %s -> %s (%s) %s", srcPort, dstPort, serviceInfo, flags)
	} else {
		return fmt.Sprintf("TCP %s -> %s %s", srcPort, dstPort, flags)
	}
}

// getTcpFlagsDesc returns a description of TCP flags
func (a *AnalyzeTCP) getTcpFlagsDesc(flags string) string {
	var descriptions []string
	
	if strings.Contains(flags, "SYN") {
		descriptions = append(descriptions, "SYN")
	}
	if strings.Contains(flags, "ACK") {
		descriptions = append(descriptions, "ACK")
	}
	if strings.Contains(flags, "FIN") {
		descriptions = append(descriptions, "FIN")
	}
	if strings.Contains(flags, "RST") {
		descriptions = append(descriptions, "RST")
	}
	if strings.Contains(flags, "PSH") {
		descriptions = append(descriptions, "PSH")
	}
	if strings.Contains(flags, "URG") {
		descriptions = append(descriptions, "URG")
	}
	if strings.Contains(flags, "ECE") {
		descriptions = append(descriptions, "ECE")
	}
	if strings.Contains(flags, "CWR") {
		descriptions = append(descriptions, "CWR")
	}
	if strings.Contains(flags, "NS") {
		descriptions = append(descriptions, "NS")
	}
	
	if len(descriptions) > 0 {
		return strings.Join(descriptions, ", ")
	}
	return "None"
}

// getServiceNameFromPort returns a service name for a given port
func (a *AnalyzeTCP) getServiceNameFromPort(port string) string {
	wellKnownPorts := map[string]string{
		"20": "FTP-DATA",
		"21": "FTP",
		"22": "SSH",
		"23": "Telnet",
		"25": "SMTP",
		"53": "DNS",
		"67": "DHCP-Server",
		"68": "DHCP-Client",
		"80": "HTTP",
		"110": "POP3",
		"143": "IMAP",
		"443": "HTTPS",
		"465": "SMTPS",
		"993": "IMAPS",
		"995": "POP3S",
		"3389": "RDP",
	}
	
	if service, ok := wellKnownPorts[port]; ok {
		return service
	}
	return ""
}

File: packet_seaquence/log/latest/work/src/protocol_analyzer/tcp.py
#!/usr/bin/env python3
"""
TCPプロトコル解析クラスを提供するモジュール
"""
from src.protocol_analyzer.base import ProtocolAnalyzer


class AnalyzeTcp(ProtocolAnalyzer):
    """
    TCP内の詳細情報を解析するクラス
    """
    
    def analyze(self):
        """
        パケットからTCP情報を解析する
        
        Returns:
            dict: TCP情報を含む辞書、または情報がない場合はNone
        """
        if not hasattr(self.packet, 'tcp'):
            return None
        
        try:
            tcp_info = {}
            
            # 送信元ポート
            if hasattr(self.packet.tcp, 'srcport'):
                tcp_info['src_port'] = self.packet.tcp.srcport
            
            # 宛先ポート
            if hasattr(self.packet.tcp, 'dstport'):
                tcp_info['dst_port'] = self.packet.tcp.dstport
            
            # シーケンス番号
            if hasattr(self.packet.tcp, 'seq'):
                tcp_info['seq'] = self.packet.tcp.seq
                if hasattr(self.packet.tcp, 'seq_raw'):
                    tcp_info['seq_raw'] = self.packet.tcp.seq_raw
            
            # 確認応答番号
            if hasattr(self.packet.tcp, 'ack'):
                tcp_info['ack'] = self.packet.tcp.ack
                if hasattr(self.packet.tcp, 'ack_raw'):
                    tcp_info['ack_raw'] = self.packet.tcp.ack_raw
            
            # フラグ
            if hasattr(self.packet.tcp, 'flags'):
                tcp_info['flags'] = self.packet.tcp.flags
                tcp_info['flags_desc'] = self._get_tcp_flags_desc(self.packet.tcp.flags)
            
            # ウィンドウサイズ
            if hasattr(self.packet.tcp, 'window_size'):
                tcp_info['window_size'] = self.packet.tcp.window_size
            elif hasattr(self.packet.tcp, 'window'):
                tcp_info['window_size'] = self.packet.tcp.window
            
            # チェックサム
            if hasattr(self.packet.tcp, 'checksum'):
                tcp_info['checksum'] = self.packet.tcp.checksum
            
            # 緊急ポインタ
            if hasattr(self.packet.tcp, 'urgent_pointer'):
                tcp_info['urgent_pointer'] = self.packet.tcp.urgent_pointer
            
            # オプション
            if hasattr(self.packet.tcp, 'options'):
                tcp_info['options'] = self.packet.tcp.options
            
            # ペイロード
            if hasattr(self.packet.tcp, 'payload'):
                tcp_info['payload'] = self.packet.tcp.payload
            elif hasattr(self.packet.tcp, 'payload_raw'):
                tcp_info['payload_raw'] = self.packet.tcp.payload_raw
            
            # ストリーム番号
            if hasattr(self.packet.tcp, 'stream'):
                tcp_info['stream'] = self.packet.tcp.stream
            
            return tcp_info
            
        except Exception as e:
            print(f"TCP解析中にエラーが発生しました: {str(e)}")
            return None
    
    def get_display_info(self):
        """
        表示用のTCP情報を取得する
        
        Returns:
            str: 表示用のTCP情報
        """
        tcp_info = self.analyze()
        if not tcp_info:
            return "TCP情報なし"
        
        src_port = tcp_info.get('src_port', '不明')
        dst_port = tcp_info.get('dst_port', '不明')
        flags = tcp_info.get('flags_desc', '不明')
        seq = tcp_info.get('seq', 'N/A')
        ack = tcp_info.get('ack', 'N/A')
        
        port_info = f"{src_port} -> {dst_port}"
        flag_info = f"[{flags}]"
        
        if 'SYN' in flags and 'ACK' not in flags:
            return f"TCP {port_info} {flag_info} 接続開始"
        elif 'SYN' in flags and 'ACK' in flags:
            return f"TCP {port_info} {flag_info} 接続確認"
        elif 'FIN' in flags:
            return f"TCP {port_info} {flag_info} 接続終了"
        elif 'RST' in flags:
            return f"TCP {port_info} {flag_info} 接続リセット"
        elif 'ACK' in flags:
            return f"TCP {port_info} {flag_info} 確認応答 (SEQ={seq}, ACK={ack})"
        else:
            return f"TCP {port_info} {flag_info}"
    
    def get_summary(self):
        """
        TCP情報のサマリーを取得する
        
        Returns:
            str: TCP情報のサマリー
        """
        tcp_info = self.analyze()
        if not tcp_info:
            return "TCP"
        
        src_port = tcp_info.get('src_port', '')
        dst_port = tcp_info.get('dst_port', '')
        flags = tcp_info.get('flags_desc', '')
        
        service_info = self._get_service_name_from_port(dst_port)
        
        if service_info:
            return f"TCP {src_port} -> {dst_port} ({service_info}) {flags}"
        else:
            return f"TCP {src_port} -> {dst_port} {flags}"
    
    def _get_tcp_flags_desc(self, flags):
        """
        TCPフラグの説明を取得する
        
        Args:
            flags (str): TCPフラグ
        
        Returns:
            str: TCPフラグの説明
        """
        descriptions = []
        if 'SYN' in flags:
            descriptions.append('SYN')
        if 'ACK' in flags:
            descriptions.append('ACK')
        if 'FIN' in flags:
            descriptions.append('FIN')
        if 'RST' in flags:
            descriptions.append('RST')
        if 'PSH' in flags:
            descriptions.append('PSH')
        if 'URG' in flags:
            descriptions.append('URG')
        if 'ECE' in flags:
            descriptions.append('ECE')
        if 'CWR' in flags:
            descriptions.append('CWR')
        if 'NS' in flags:
            descriptions.append('NS')
        
        return ', '.join(descriptions) if descriptions else 'None'
    
    def _get_service_name_from_port(self, port):
        """
        ポート番号からサービス名を取得する
        
        Args:
            port (str): ポート番号
        
        Returns:
            str: サービス名、または空文字列
        """
        try:
            port_int = int(port)
            well_known_ports = {
                20: 'FTP-DATA',
                21: 'FTP',
                22: 'SSH',
                23: 'Telnet',
                25: 'SMTP',
                53: 'DNS',
                67: 'DHCP-Server',
                68: 'DHCP-Client',
                80: 'HTTP',
                110: 'POP3',
                143: 'IMAP',
                443: 'HTTPS',
                465: 'SMTPS',
                993: 'IMAPS',
                995: 'POP3S',
                3389: 'RDP'
            }
            return well_known_ports.get(port_int, '')
        except (ValueError, TypeError):
            return ''
